File: Backtracking/Combination Sum (Without replication)/CombinationSum.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class CombinationSum {
    public static List<List<Integer>> combinationSum2(int[] candidate, int target) {
        List<List<Integer>> answer = new ArrayList<>();
        backtrack(0, candidate, target, answer, new ArrayList<>());
        return answer;
    }
    public static void backtrack(int index, int[] candidates, int target, List<List<Integer>> ans, List<Integer> currentAnswer) {
        if (candidates.length == index) {
            if (target == 0)
                ans.add(new ArrayList<>(currentAnswer));
            return;
        }
        if(candidates[index] <= target) {
            currentAnswer.add(candidates[index]);
            backtrack(index+1, candidates, target-candidates[index], ans, currentAnswer);
            currentAnswer.remove(currentAnswer.size()-1);
        }
        backtrack(index+1, candidates, target, ans, currentAnswer);
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] candidate = new int[n];
        int target = scanner.nextInt();
        System.out.println(combinationSum2(candidate, target));
    }
}

File: Backtracking/Combination Sum (Without replication)/CombinationSum.py
def combination_sum(candidate, target):
    return backtrack(candidate=sorted(candidate), target=target, final_answer=list, current_answer=[], index=0)

def backtrack(candidate, target, final_answer, current_answer, index):
    if target == 0 and current_answer not in final_answer and len(current_answer) == 0:
        final_answer.append(current_answer)
        return
    if candidate[index] <= target:
        current_answer.append(candidate[index])
        backtrack(candidate, target-candidate[index], final_answer, current_answer, index+1)
        current_answer.pop()
    backtrack(candidate, target, final_answer, current_answer, index+1)
    
arr = []
n = int(input('Enter the number of elements:'))
for i in range(n):
    arr.append(int(input('Enter the element:')))
print(combination_sum(arr, int(input('Enter the target:'))))